core/sys/isrControl.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2015, Scanimetrics - http://www.scanimetrics.com
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/*---------------------------------------------------------------------------*/
+#ifndef INTERRUPTS_H_
+#define INTERRUPTS_H_
+/*---------------------------------------------------------------------------*/
+/**
+* Includes
+**/
+#include "cc.h"
+#include "isrControlArch.h"
+/*---------------------------------------------------------------------------*/
+/**
+ * Disables global interrupts, returns true if interrupts were already disabled
+**/
+static CC_INLINE int _arch_disableINTS();
+/**
+ * Enables global interrupts
+**/
+static CC_INLINE void _arch_enableINTS();
+/*---------------------------------------------------------------------------*/
+/**
+* Macros
+**/
+#define disableINTS _arch_disableINTS
+#define enableINTS  _arch_enableINTS
+
+#define ATOMIC_SECTION_ENTER \
+    {int _ints_disabled = _arch_disableINTS();
+
+#define ATOMIC_SECTION_EXIT \
+    if(!_ints_disabled) _arch_enableINTS();}
+/*---------------------------------------------------------------------------*/
```

Comment:
@tonnenpinguin 

ATOMIC_SECTION_ENTER starts a block scope which ATOMIC_SECTION_EXIT ends. This will do two things:
1. You will not be able to use ATOMIC_SECTION_ENTER in a different scope than ATOMIC_SECTION_EXIT (the compiler will complain about an unmatched curly brace)
2. Even old compilers will allow the variable definition since the variable is defined at the start of a new scope

With that said, this still has its problems but I've chosen to implement it this way because this is already a fairly common idiom IMO (see for example http://www.keil.com/forum/20720/atomic-sections-of-code/)